### ISSUE TITLE
Implicit imports break code linting

### ISSUE THREAD

**Issue by fabianegli** [OpenMS/OpenMS#6338]
The implicit imports break code linting tools. Could this be changed to explicit imports?

https://github.com/OpenMS/OpenMS/blob/e4143aacf975dc1f7017aa91edb82c2ce95cd3ac/src/pyOpenMS/pyopenms/__init__.py#L65-L66

**Comment by timosachsenberg** [comment]
Not a python export here but I think we can't add explicit imports here because at this stage we did not generate the python classes, yet.

**Comment by fabianegli** [comment]
I am not familiar with the use of C++ extensions and don't know either. Could you try and find out if it works with an explicit import? e.g. try adding `from .all_modules import AASequence` as an experiment above the implicit imports.

**Comment by jpfeuffer** [comment]
I agree with @timosachsenberg that listing ALL classes is not a scalable approach. I think we could offer adding "ignore" comments for the most common linters.

Middle-term we should implement a solution for correctly merging multiple C extensions into one module:
https://stackoverflow.com/questions/30157363/collapse-multiple-submodules-to-one-cython-extension/52729181#52729181
https://docs.cython.org/en/latest/src/userguide/source_files_and_compilation.html#integrating-multiple-modules

Related issues for API docs: https://github.com/OpenMS/pyopenms-docs/issues/264

**Comment by fabianegli** [comment]
I don't quite follow why this is not scalable - or why this would matter much here.

I would not start ignoring things - the linting to date isn't (expected to) work and I can readily ignore the errors I see and would prefer to no see this situation be baked into the codebase.

+1 fot the middle term planning of the cython modules collapse.

There's also a question if you'd like to adopt a specific api submodule as e.g. in astropy or statsmodels.

Another option might be to define the special variable `__all__` to also hide part of the things available that don't need to be.

**Comment by jpfeuffer** [comment]
The problem is that we cannot know in which submodule a class ends up. We have a splitting mechanism to parallelize Cython compilation. It is purely based on equal splitting of the number of classes.
And it is a bit counterintuitive for a contributor to also touch that file, whenever a class is introduced. Plus a burden during code review, to always keep an eye on that.

I think ignoring is a viable solution here. We know that we always want all classes of the split submodules to be exposed in the parent module without specifying the submodule number where it comes from. `*` is a totally valid solution in this case IMO.

Yes, context-specific splitting would be great. E.g. creating submodules based on the C++ source folders until we have a complete modularization into multiple C++ sublibraries that we can wrap (which will be a much bigger task, see https://github.com/OpenMS/OpenMS/pull/5613 for a first attempt).

I don't think we need `__all__` since we decide through the pxd files what is supposed to be exposed to Python.

**Comment by fabianegli** [comment]
> [*] import(s) os, here, sys, ctypes, e, PyQt4, info, c, libc, SysInfo, free_mem, win32api, SimpleOpenMSSpectraFactory, Constants and Plot from wildcard import of pyopenms

Some of them might be reasonable to omit when crafting the contents of `__all__`

**Comment by fabianegli** [comment]
``

> The problem is that we cannot know in which submodule a class ends up. We have a splitting mechanism to parallelize Cython compilation. It is purely based on equal splitting of the number of classes.

Then maybe the imports could be rewritten programmatically as part of the preparations for the compilation?

**Comment by jpfeuffer** [comment]
> > [*] import(s) os, here, sys, ctypes, e, PyQt4, info, c, libc, SysInfo, free_mem, win32api, SimpleOpenMSSpectraFactory, Constants and Plot from wildcard import of pyopenms
> 
> Some of them might be reasonable to omit when crafting the contents of `__all__`

Maybe we shouldn't star import there, then.


> Then maybe the imports could be rewritten programmatically as part of the preparations for the compilation?

Yes, this is most probably possible but unfortunately, linting does not bug me enough to spend my time on doing this. I am happy to give hints if anyone wants to do it.

**Comment by jpfeuffer** [comment]
By the way, which linter are you using? Good linters should create an exception for wildcard imports in `__init__.py`.
According to PEP 8: "There is one defensible use case for a wildcard import, which is to republish an internal interface as part of a public API." which is what we are doing.
See e.g. sonarcloud https://rules.sonarsource.com/python/RSPEC-2208

**Comment by jpfeuffer** [comment]
maybe #6364 reduces the impact and/or makes it clearer that we are just reexporting an internal API.
